python/ray/autoscaler/aws/config.py

```diff
@@ -319,6 +294,39 @@ def _check_ami(config):
                         region=region))
 
 
+def _upsert_security_groups(config):
+    group_name = SECURITY_GROUP_TEMPLATE.format(config["cluster_name"])
+
+    security_groups = _get_or_create_vpc_security_groups(config, group_name)
+    _upsert_security_group_rules(config, security_groups)
+
+    return security_groups
+
+
+def _get_or_create_vpc_security_groups(conf, group_name):
+    worker_subnet_id = conf["worker_nodes"]["SubnetIds"][0]
+    head_subnet_id = conf["head_node"]["SubnetIds"][0]
+
+    worker_vpc_id = _get_vpc_id_or_die(conf, worker_subnet_id)
+    head_vpc_id = _get_vpc_id_or_die(conf, head_subnet_id) \
+        if head_subnet_id != worker_subnet_id else worker_vpc_id
+
+    vpc_ids = [worker_vpc_id, head_vpc_id] \
+        if head_vpc_id != worker_vpc_id else [worker_vpc_id]
+    all_sgs = _get_security_groups(conf, vpc_ids, [group_name])
+
+    worker_sg = next((_ for _ in all_sgs if _.vpc_id == worker_vpc_id), None)
+    if worker_sg is None:
```

Comment:
Since we only enter this path if a user didn't explicitly specify head and worker security group IDs, we know here that we have to create at least one security group using the default group name passed into this method.

However (and this may be what you're pointing out), a user could specify an explicit security group ID for either head or worker nodes, but not both. If a user then also specifies an explicit subnet that puts head and worker nodes on different VPCs, then the current code would create an unused default security group on one of these VPCs.

Since we're largely just repeating the same operation for head and worker nodes here and elsewhere, one way to both resolve this issue and simplify the code is to pass in the list of candidate node types (e.g. ["head", "workers"]) to consider for default security group configuration (i.e. the node types that don't have an explicitly configured security group ID).

I'm thinking something like this:

``` python
NODE_TYPE_CONFIG_KEYS = {
    "workers": "worker_nodes",
    "head": "head_node"
}

def _configure_security_group(config):
    node_types_to_configure = [
        node_type for node_type, config_key in NODE_TYPE_CONFIG_KEYS.items()
        if "SecurityGroupIds" not in config[NODE_TYPE_CONFIG_KEYS[node_type]]
    ]
    if not node_types_to_configure:
        return config  # have user-defined groups

    security_groups = _upsert_security_groups(config, node_types_to_configure)
    ########################################################
    # remaining implementation logic intentionally omitted #
    ########################################################

def _get_or_create_vpc_security_groups(conf, group_name, node_types):
    subnet_to_vpc, node_type_to_vpc, node_type_to_sg = {}, {}, {}
    node_type_to_subnet = {
        node_type: conf[NODE_TYPE_CONFIG_KEYS[node_type]]["SubnetIds"][0]
        for node_type in node_types
    }
    for node_type in node_types:
        subnet_id = node_type_to_subnet[node_type]
        subnet_to_vpc.setdefault(
            subnet_id,
            _get_vpc_id_or_die(conf, node_type_to_subnet[node_type]))
        node_type_to_vpc[node_type] = subnet_to_vpc[subnet_id]

    all_sgs = _get_security_groups(conf, subnet_to_vpc.values(), [group_name])
    for node_type in node_types:
        vpc_id = node_type_to_vpc[node_type]
        sg = next((_ for _ in all_sgs if _.vpc_id == vpc_id), None)
        if sg is None:
            sg = _create_security_group(conf, vpc_id, group_name)
            all_sgs.append(sg)
        node_type_to_sg[node_type] = sg

    return node_type_to_sg
```

As far as invalid explicit security group ID specification goes, the historic behavior in this code path has been to simply accept any security group ID that a user puts in without ensuring its existence. This error will be caught further downstream in the cluster configuration process with a small net benefit of slightly lower latency during config bootstrapping. I have mixed feelings about this behavior, but feel OK about it, since I recall the downstream error arising pretty quickly and clearly pointing out that the specified security group doesn't exist.